Adds an aten::_ops namespace with unambiguous function names

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58092 Adds an aten::_ops namespace with unambiguous function names**

Fixes #58044.

This PR:
- adds `ATEN_FN(op)` and `ATEN_FN2(op, overload)` macros that resolve to
an non-overloaded function in aten::_ops that calls the desired operator
(without default arguments).

The motivation for this is two-fold:
1) Using aten operators with templates is hard if the operator is
overloaded (e.g. add.Tensor and add.Scalar).
2) Method-only operators require special handling; pointers-to-method
are different from function pointers. `ATEN_FN2(add_, Tensor)` returns
a function instead of a method.

There is some interesting behavior for out= operations.
`ATEN_FN2(sin, "out")` gives a function that is *faithful* to the schema;
that is, the order of arguments is exactly what it looks like in the
schema. This makes it so that you can directly register
`ATEN_FN2(sin,"out")` (or a function wrapping it using the same signature)
as an override for a DispatchKey.

Test Plan:
- New tests that ATEN_FN2 works on function and method-only operators
- New test that ATEN_FN works
- New test that ATEN_FN macro returns a "faithful" function.

Codegen output:
Operators.h and Operators.cpp are both here:
https://gist.github.com/zou3519/c2c6a900410b571f0d7d127019ca5175

Differential Revision: [D28643215](https://our.internmc.facebook.com/intern/diff/D28643215)

zou3519 committed:
Adds an aten::_ops namespace with unambiguous function names

Fixes #58044.

This PR:
- adds `ATEN_FN(op)` and `ATEN_FN2(op, overload)` macros that resolve to
an non-overloaded function in aten::_ops that calls the desired operator
(without default arguments).

The motivation for this is two-fold:
1) Using aten operators with templates is hard if the operator is
overloaded (e.g. add.Tensor and add.Scalar).
2) Method-only operators require special handling; pointers-to-method
are different from function pointers. `ATEN_FN2(add_, Tensor)` returns
a function instead of a method.

There is some interesting behavior for out= operations.
`ATEN_FN2(sin, "out")` gives a function that is *faithful* to the schema;
that is, the order of arguments is exactly what it looks like in the
schema. This makes it so that you can directly register
`ATEN_FN2(sin,"out")` (or a function wrapping it using the same signature)
as an override for a DispatchKey.

Test Plan:
- New tests that ATEN_FN2 works on function and method-only operators
- New test that ATEN_FN works
- New test that ATEN_FN macro returns a "faithful" function.

[ghstack-poisoned]

zou3519 committed:
Update on "Adds an aten::_ops namespace with unambiguous function names"


Fixes #58044.

This PR:
- adds `ATEN_FN(op)` and `ATEN_FN2(op, overload)` macros that resolve to
an non-overloaded function in aten::_ops that calls the desired operator
(without default arguments).

The motivation for this is two-fold:
1) Using aten operators with templates is hard if the operator is
overloaded (e.g. add.Tensor and add.Scalar).
2) Method-only operators require special handling; pointers-to-method
are different from function pointers. `ATEN_FN2(add_, Tensor)` returns
a function instead of a method.

There is some interesting behavior for out= operations.
`ATEN_FN2(sin, "out")` gives a function that is *faithful* to the schema;
that is, the order of arguments is exactly what it looks like in the
schema. This makes it so that you can directly register
`ATEN_FN2(sin,"out")` (or a function wrapping it using the same signature)
as an override for a DispatchKey.

Test Plan:
- New tests that ATEN_FN2 works on function and method-only operators
- New test that ATEN_FN works
- New test that ATEN_FN macro returns a "faithful" function.

Codegen output:
-
[UnambiguousTorchOperators.h](https://gist.github.com/zou3519/8be136b37ebcd0551eb63d9e01d39e2b)
-
[UnambiguousTorchOperators.cpp](https://gist.github.com/zou3519/7a6d74e2ca04daef17cecbf2999a6fb7)

Some interesting things to look at:
- [the mul
overloads](https://gist.github.com/zou3519/7a6d74e2ca04daef17cecbf2999a6fb7#file-unambiguoustorchoperators-cpp-L1957-L1971)

[ghstack-poisoned]

zou3519 committed:
Update on "Adds an aten::_ops namespace with unambiguous function names"


Fixes #58044.

This PR:
- adds `ATEN_FN(op)` and `ATEN_FN2(op, overload)` macros that resolve to
an non-overloaded function in aten::_ops that calls the desired operator
(without default arguments).

The motivation for this is two-fold:
1) Using aten operators with templates is hard if the operator is
overloaded (e.g. add.Tensor and add.Scalar).
2) Method-only operators require special handling; pointers-to-method
are different from function pointers. `ATEN_FN2(add_, Tensor)` returns
a function instead of a method.

There is some interesting behavior for out= operations.
`ATEN_FN2(sin, "out")` gives a function that is *faithful* to the schema;
that is, the order of arguments is exactly what it looks like in the
schema. This makes it so that you can directly register
`ATEN_FN2(sin,"out")` (or a function wrapping it using the same signature)
as an override for a DispatchKey.

Test Plan:
- New tests that ATEN_FN2 works on function and method-only operators
- New test that ATEN_FN works
- New test that ATEN_FN macro returns a "faithful" function.

Codegen output:
-
[UnambiguousTorchOperators.h](https://gist.github.com/zou3519/8be136b37ebcd0551eb63d9e01d39e2b)
-
[UnambiguousTorchOperators.cpp](https://gist.github.com/zou3519/7a6d74e2ca04daef17cecbf2999a6fb7)

Some interesting things to look at:
- [the mul
overloads](https://gist.github.com/zou3519/7a6d74e2ca04daef17cecbf2999a6fb7#file-unambiguoustorchoperators-cpp-L1957-L1971)

[ghstack-poisoned]

zou3519 committed:
Update on "Adds an aten::_ops namespace with unambiguous function names"


Fixes #58044.

This PR:
- adds `ATEN_FN(op)` and `ATEN_FN2(op, overload)` macros that resolve to
an non-overloaded function in aten::_ops that calls the desired operator
(without default arguments).

The motivation for this is two-fold:
1) Using aten operators with templates is hard if the operator is
overloaded (e.g. add.Tensor and add.Scalar).
2) Method-only operators require special handling; pointers-to-method
are different from function pointers. `ATEN_FN2(add_, Tensor)` returns
a function instead of a method.

There is some interesting behavior for out= operations.
`ATEN_FN2(sin, "out")` gives a function that is *faithful* to the schema;
that is, the order of arguments is exactly what it looks like in the
schema. This makes it so that you can directly register
`ATEN_FN2(sin,"out")` (or a function wrapping it using the same signature)
as an override for a DispatchKey.

Test Plan:
- New tests that ATEN_FN2 works on function and method-only operators
- New test that ATEN_FN works
- New test that ATEN_FN macro returns a "faithful" function.

Codegen output:
- [UnambiguousTorchOperators.h](https://gist.github.com/zou3519/8be136b37ebcd0551eb63d9e01d39e2b)
- [UnambiguousTorchOperators.cpp](https://gist.github.com/zou3519/7a6d74e2ca04daef17cecbf2999a6fb7)

Some interesting things to look at:
- [the mul
overloads](https://gist.github.com/zou3519/7a6d74e2ca04daef17cecbf2999a6fb7#file-unambiguoustorchoperators-cpp-L1957-L1971)

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/templates/UnambiguousTorchOperators.h:
```diff
@@ -0,0 +1,41 @@
+
+// NB: this is special cased here because requires_grad_ returns a
+// const Tensor&. The codegen assumes that in-place functions return Tensor&.
+TORCH_API Tensor & requires_grad_(Tensor & self, bool requires_grad);
```
What happens when you let the codegen handle `requires_grad_`? The comment says that it returns a const Tensor& but the declaration here just returns a normal Tensor&.

zou3519 replied:
If I let the codegen handle `requires_grad_`, it outputs the following definition:
```
Tensor & requires_grad_(Tensor & self, bool requires_grad) {
    return self.requires_grad_(requires_grad);
}
```
Now, this is problematic because `self.requires_grad_(requires_grad)` returns a `const Tensor&`, but `ATEN_FN(requires_grad_)` returns a Tensor&, so the compiler complains that we're coercing a `const Tensor&` to a `Tensor&`. All other in-place functions return non-const references (e.g. `self.sin_()` returns a Tensor&).

`ATEN_FN(requires_grad_)` should probably have a signature that is equal to whatever signature an override for `requires_grad_` should be. build/aten/src/ATen/core/TensorMethods.cpp shows that that signature should be `at::Tensor & (at::Tensor &, bool)`:
```
// aten::requires_grad_(Tensor(a!) self, bool requires_grad=True) -> Tensor(a!)
at::Tensor & Tensor::__dispatch_requires_grad_(bool requires_grad) const {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::requires_grad_", "")
        .typed<at::Tensor & (at::Tensor &, bool)>();
    return op.call(const_cast<Tensor&>(*this), requires_grad);
}
```

bdhirsh replied:
hmmm ok, so it sounds like all of the auto-generated `inplace` method signatures in `TensorMethods.h` are codegen'd to return `Tensor&`, but you're forced to special case `requires_grad_` because it has a `manual_cpp_binding` with a [different signature](https://github.com/pytorch/pytorch/blob/5ea87f9c24cdc7bfde9e4cacca1985592d65e702/aten/src/ATen/templates/TensorBody.h#L868) that returns `const Tensor&`.

I'm pretty sure that long term, we want to eventually kill `Tensor&` from any C++ signatures and use `const Tensor&` everywhere, but it'd take some work to get the codegen to uniformly do this.

Maybe just include somewhere in the comment that we eventually want to remove that special case and let the codegen handle it, once we remove mutable `Tensor&` returns from the codegen?

bdhirsh commented on aten/src/ATen/templates/UnambiguousTorchOperators.h:
```diff
@@ -0,0 +1,41 @@
+// Use ATEN_FN(op) to get an unambiguous *function* version of the operator.
+//
+// There is some interesting behavior for out= operations.
+// ATEN_FN2(sin, "out") gives a function that is *faithful* to the schema;
```
nit: are there supposed to be quotes around `out`?

zou3519 replied:
Good catch!

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -176,6 +176,94 @@ def __call__(self, f: NativeFunction) -> Optional[str]:
+        unambiguous_name = self.unambiguous_function_name(f)
+        # NB: requires_grad is the only exception to the rule because
+        # its const correctness is questionable.
+        if unambiguous_name in set(['requires_grad_']):
```
most of the other parsing in the codegen works directly off of the NativeFunction object instead of some other converted value. So you don't really need to call `self.unambiguous_function_name()` to do the comparison here; you can just do `if str(f.func.name) in set(['requires_grad_'])`

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -176,6 +176,94 @@ def __call__(self, f: NativeFunction) -> Optional[str]:
+
+    def gen_declaration(self, f: NativeFunction) -> str:
+        unambiguous_name = self.unambiguous_function_name(f)
+        returns_type = dispatcher.returns_type(f.func.returns).cpp_type_registration_declarations()
```
You can use the signature classes in `types.py`, plus `translate()` to do most of the work for you when generating the declaration/definitions and translating from the dispatcher API to C++ API. Taking some inspiration from [`register_dispatch_key.py`](https://github.com/pytorch/pytorch/blob/ff982ef73d91bf384d8f6df65a0a63dfbc998b7d/tools/codegen/dest/register_dispatch_key.py#L149), something like:

```
unambiguous_name = self.unambiguous_function_name(f)
sig = DispatcherSignature.from_schema(f)

if self.target is Target.Declaration:
    return f'TORCH_API {sig.decl(unambiguous_name)};'
elif self.target is Target.Definition:
    target_sig = CppSignatureGroup.from_native_function(f, method=False).most_faithful_signature()
    return f"""
{sig.defn(unambiguous_name)} {{
    return {target_sig.name()}({', '.join(e.expr for e in translate(sig.arguments(), target_sig.arguments()))});
}}"""
    else:
        assert_never(self.target)
```

I'm pretty sure that `CppSignatureGroup.from_native_function(f, method=False)` should do all the work of dealing with method-only ops for you.

zou3519 replied:
Thanks for the tip! DispatcherSignature.decl()/defn() work and make this a lot simpler.

Is there a utility function somewhere that handles writing invocations for method-only operators (e.g. "at::sin(self)" vs "self.sin_()")? I couldn't find one so I rolled my own in the updated code

zou3519 committed:
Update on "Adds an aten::_ops namespace with unambiguous function names"


Fixes #58044.

This PR:
- adds `ATEN_FN(op)` and `ATEN_FN2(op, overload)` macros that resolve to
an non-overloaded function in aten::_ops that calls the desired operator
(without default arguments).

The motivation for this is two-fold:
1) Using aten operators with templates is hard if the operator is
overloaded (e.g. add.Tensor and add.Scalar).
2) Method-only operators require special handling; pointers-to-method
are different from function pointers. `ATEN_FN2(add_, Tensor)` returns
a function instead of a method.

There is some interesting behavior for out= operations.
`ATEN_FN2(sin, "out")` gives a function that is *faithful* to the schema;
that is, the order of arguments is exactly what it looks like in the
schema. This makes it so that you can directly register
`ATEN_FN2(sin,"out")` (or a function wrapping it using the same signature)
as an override for a DispatchKey.

Test Plan:
- New tests that ATEN_FN2 works on function and method-only operators
- New test that ATEN_FN works
- New test that ATEN_FN macro returns a "faithful" function.

Codegen output:
- [UnambiguousTorchOperators.h](https://gist.github.com/zou3519/8be136b37ebcd0551eb63d9e01d39e2b)
- [UnambiguousTorchOperators.cpp](https://gist.github.com/zou3519/7a6d74e2ca04daef17cecbf2999a6fb7)

Some interesting things to look at:
- [the mul
overloads](https://gist.github.com/zou3519/7a6d74e2ca04daef17cecbf2999a6fb7#file-unambiguoustorchoperators-cpp-L1957-L1971)

[ghstack-poisoned]

ezyang commented on tools/codegen/gen.py:
```diff
@@ -176,6 +176,85 @@ def __call__(self, f: NativeFunction) -> Optional[str]:
+
+    # Given a function schema "aten::op.overload(...)",
+    # If there is no overload name, this returns f"{op}"
+    # If there is an overload name, this returns f"{op}_{overload}"
```
worth commenting that this must be synchronized with ATEN_FN/ATEN_FN2

ezyang commented:
> Is there a utility function somewhere that handles writing invocations for method-only operators (e.g. "at::sin(self)" vs "self.sin_()")? I couldn't find one so I rolled my own in the updated code

I don't know how much time you want to spend on this PR, but if you're willing to do more refactoring, what might end up working out better in the end is nuking ComputeFunction/ComputeTensorMethod.  The idea is that now Unambiguous is the canonical definition (with the dispatcher code), and then ComputeFunction/ComputeTensorMethod call into *this*. (I might also think about coming up with a better name than Unambiguous; maybe just ComputeOp in symmetry with how all dispatcher register functions are available at torch.ops)

zou3519 commented:
> I don't know how much time you want to spend on this PR, but if you're willing to do more refactoring, what might end up working out better in the end is nuking ComputeFunction/ComputeTensorMethod. The idea is that now Unambiguous is the canonical definition (with the dispatcher code), and then ComputeFunction/ComputeTensorMethod call into this.

I'm a bit confused about what the suggestion is. ComputeFunction/ComputeTensorMethod generate calls to the dispatcher that look like:
```
// aten::mul.Tensor(Tensor self, Tensor other) -> Tensor
at::Tensor mul(const at::Tensor & self, const at::Tensor & other) {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::mul", "Tensor")
        .typed<at::Tensor (const at::Tensor &, const at::Tensor &)>();
    return op.call(self, other);
}
// aten::mul_.Tensor(Tensor(a!) self, Tensor other) -> Tensor(a!)
at::Tensor & Tensor::mul_(const at::Tensor & other) const {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::mul_", "Tensor")
        .typed<at::Tensor & (at::Tensor &, const at::Tensor &)>();
    return op.call(const_cast<Tensor&>(*this), other);
}
```
while UnambiguousTorchOps generates code that looks like:
```
Tensor mul_Tensor(const Tensor & self, const Tensor & other) {
  return at::mul(self, other);
}        
Tensor & mul__Tensor(Tensor & self, const Tensor & other) {
  return self.mul_(other);
}  
```
Is the proposal to change UnambiguousTorchOps to actually do the dispatches and ComputeFunction/ComputeTensorMethod to use UnambiguousTorchOps?
```
// UnambiguousTorchOps
Tensor mul_Tensor(const Tensor & self, const Tensor & other) {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::mul", "Tensor")
        .typed<at::Tensor (const at::Tensor &, const at::Tensor &)>();
    return op.call(self, other);
}        
Tensor & mul__Tensor(Tensor & self, const Tensor & other) {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::mul_", "Tensor")
        .typed<at::Tensor & (at::Tensor &, const at::Tensor &)>();
    return op.call(const_cast<Tensor&>(*this), other);
}  
// Functions.cpp
// aten::mul.Tensor(Tensor self, Tensor other) -> Tensor
at::Tensor mul(const at::Tensor & self, const at::Tensor & other) {
    return ATEN_FN(mul, Tensor)(self, other);
}
// TensorMethods.cpp
at::Tensor & Tensor::mul_(const at::Tensor & other) const {
    return ATEN_FN(mul_, Tensor)(*this, other);
}
```
> (I might also think about coming up with a better name than Unambiguous; maybe just ComputeOp in symmetry with how all dispatcher register functions are available at torch.ops)

Agree that Unambiguous is not the best name

ezyang commented:
> Is the proposal to change UnambiguousTorchOps to actually do the dispatches and ComputeFunction/ComputeTensorMethod to use UnambiguousTorchOps?

yup

zou3519 commented:
Can I get a hot take:

ComputeFunctions generates not only Functions.{h, cpp}, but also RedispatchingFunctions.{h, cpp}. ComputeTensorMethods generates TensorMethods.{h, cpp}, but there isn't a "redispatching" version of it.

Right now the proposed refactor is to have ATEN_FN2/ATEN_FN be responsible for dispatching and have Functions.{h, cpp} and TensorMethods.{h, cpp} have functions/methods that are thin wrappers around ATEN_FN2/ATEN_FN. 

I'm not planning on changing RedispatchingFunctions.{h, cpp}, but I wanted to check in: is there any reason to drag RedispatchingFunctions.{h, cpp} into this refactor as well? That would involve "redispatching" variants of the ATEN_FN2/ATEN_FN macros, but I don't see any use cases for that.

bdhirsh commented:
> Can I get a hot take:
> 
> ComputeFunctions generates not only Functions.{h, cpp}, but also RedispatchingFunctions.{h, cpp}. ComputeTensorMethods generates TensorMethods.{h, cpp}, but there isn't a "redispatching" version of it.
> 
> Right now the proposed refactor is to have ATEN_FN2/ATEN_FN be responsible for dispatching and have Functions.{h, cpp} and TensorMethods.{h, cpp} have functions/methods that are thin wrappers around ATEN_FN2/ATEN_FN.
> 
> I'm not planning on changing RedispatchingFunctions.{h, cpp}, but I wanted to check in: is there any reason to drag RedispatchingFunctions.{h, cpp} into this refactor as well? That would involve "redispatching" variants of the ATEN_FN2/ATEN_FN macros, but I don't see any use cases for that.

My two cents: Agreed, separating the redispatch API into two pieces does kind of feel like overkill, since I don't really see a use for the redispatch version of `at::_ops`. I'd tentatively vote to leave the redispatch API out of the refactor, but maybe make an effort to factor out some of the shared code between the redispatch API and UnambiguousTorchOps. Stuff like the code template is really similar between the two, just with a slightly different call to the dispatcher API.

> ComputeFunctions generates not only Functions.{h, cpp}, ..., ComputeTensorMethods generates TensorMethods.{h, cpp}

One thing we also probably want to be careful of is inlining: `Functions.cpp` is a hot-path piece of code and we want to avoid an extra function boundary from `at::add` to `at::_ops::add_Tensor` if we can avoid it. Probably by just stuffing the short function definitions directly into `Functions.h`?

ezyang commented:
I have a different hot take, which is that `aten::_ops` redispatching API should subsume the old redispatching API entirely. The reason I believe this is this condition:

```
        # We unconditionally generate function variants of the redispatch API.
        # This is mainly because we can namespace functions separately, but not methods,
        if Variant.function not in f.variants and not self.is_redispatching_fn:
            return None
```

This was always kind of weird bifurcation of behavior between the is redispatching and not case. With @zou3519's change, there is no problem anymore: `aten::_ops` is always functions and never methods, you have an easy, uniform access pattern.

ezyang commented:
More unintentional impact: with the unambiguous namespace https://github.com/pytorch/pytorch/blob/2ddd8416350b701fe0e7325e4761f8d14a3f46e5/aten/src/ATen/autocast_mode.cpp#L261 can stop repeating all the type signatures

zou3519 committed:
Update on "Adds an aten::_ops namespace with unambiguous function names"


Fixes #58044.

This PR:
- adds `ATEN_FN(op)` and `ATEN_FN2(op, overload)` macros that resolve to
an non-overloaded function in aten::_ops that calls the desired operator
(without default arguments).

The motivation for this is two-fold:
1) Using aten operators with templates is hard if the operator is
overloaded (e.g. add.Tensor and add.Scalar).
2) Method-only operators require special handling; pointers-to-method
are different from function pointers. `ATEN_FN2(add_, Tensor)` returns
a function instead of a method.

There is some interesting behavior for out= operations.
`ATEN_FN2(sin, "out")` gives a function that is *faithful* to the schema;
that is, the order of arguments is exactly what it looks like in the
schema. This makes it so that you can directly register
`ATEN_FN2(sin,"out")` (or a function wrapping it using the same signature)
as an override for a DispatchKey.

Test Plan:
- New tests that ATEN_FN2 works on function and method-only operators
- New test that ATEN_FN works
- New test that ATEN_FN macro returns a "faithful" function.

Codegen output:
Operators.h and Operators.cpp are both here:
https://gist.github.com/zou3519/c2c6a900410b571f0d7d127019ca5175

[ghstack-poisoned]

zou3519 commented:
> So other than the smaller pieces of feedback in the PR (renaming UnambiguousOperators)

I renamed UnambiguousOperators to Operators. The include is `<ATen/Operators.h>`. This name is nice because it gives some sort of parity between the torch.ops Python namespace, but I can see it causing confusion because of how generic the name is.

Please let me know what you folks think about the new name.

bdhirsh reviewed (approved):
Operators.h/cpp seems good to me!

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28643215).

zou3519 committed:
Update on "Adds an aten::_ops namespace with unambiguous function names"


Fixes #58044.

This PR:
- adds `ATEN_FN(op)` and `ATEN_FN2(op, overload)` macros that resolve to
an non-overloaded function in aten::_ops that calls the desired operator
(without default arguments).

The motivation for this is two-fold:
1) Using aten operators with templates is hard if the operator is
overloaded (e.g. add.Tensor and add.Scalar).
2) Method-only operators require special handling; pointers-to-method
are different from function pointers. `ATEN_FN2(add_, Tensor)` returns
a function instead of a method.

There is some interesting behavior for out= operations.
`ATEN_FN2(sin, "out")` gives a function that is *faithful* to the schema;
that is, the order of arguments is exactly what it looks like in the
schema. This makes it so that you can directly register
`ATEN_FN2(sin,"out")` (or a function wrapping it using the same signature)
as an override for a DispatchKey.

Test Plan:
- New tests that ATEN_FN2 works on function and method-only operators
- New test that ATEN_FN works
- New test that ATEN_FN macro returns a "faithful" function.

Codegen output:
Operators.h and Operators.cpp are both here:
https://gist.github.com/zou3519/c2c6a900410b571f0d7d127019ca5175

Differential Revision: [D28643215](https://our.internmc.facebook.com/intern/diff/D28643215)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28643215).
